sklearn/preprocessing/_data.py

```diff
@@ -888,13 +891,13 @@ def partial_fit(self, X, y=None, sample_weight=None):
         # See incr_mean_variance_axis and _incremental_mean_variance_axis
 
         # if n_samples_seen_ is an integer (i.e. no missing values), we need to
-        # transform it to a NumPy array of shape (n_features,) required by
+        # transform it to an array of shape (n_features,) required by
         # incr_mean_variance_axis and _incremental_variance_axis
-        dtype = np.int64 if sample_weight is None else X.dtype
+        dtype = counter_dtype(xp, X.dtype) if sample_weight is None else X.dtype
```

Comment:
This was required for torch.float16 (and maybe torch.float32 as well) as the counters will be multiplied internally and will cast the internally used float16 / float32 dtype to some larger dtype, which results in an error later.

Reply:
I don't see the point of making the range of the int counter depend on the floating point precision of `X.dtype`. I think we should always use `xp.int64` to be able to support large values in `n_samples_seen_`.

Reply:
I changed it because it will have an influence on the dtype here: https://github.com/scikit-learn/scikit-learn/blob/3d9293ad18f2e7c320ea37fa0894865b9621327e/sklearn/utils/extmath.py#L1153

When there is a cast to another dtype, this will result in an error later in the code.

Reply:
We should lazily cast the integer counts to `xp.float64` whenever dividing them (or multiplying them with floating point values). For instance:

```python
    last_over_new_count = (
        xp.astype(last_sample_count, xp.float64)
       / xp.astype(new_sample_count, xp.float64)
    )
```